test/e2e/scheduling/taint_based_evictions.go

```diff
@@ -0,0 +1,188 @@
+						Operator:          v1.TolerationOpExists,
+						Effect:            v1.TaintEffectNoExecute,
+						TolerationSeconds: &tolerationSeconds[i],
+					},
+				},
+				DeletionGracePeriodSeconds: &zero,
+			})
+		}
+		// build pod2
+		pods[numPods-1] = createPausePod(f, pausePodConfig{
+			Name:     fmt.Sprintf("%v-%v", base, numPods-1),
+			NodeName: nodeName,
+		})
+
+		By("Verifying all pods are running properly")
+		for _, pod := range pods {
+			framework.ExpectNoError(framework.WaitForPodRunningInNamespace(cs, pod))
+		}
+
+		// get the node API object
+		nodeSelector := fields.OneTermEqualSelector("metadata.name", nodeName)
+		nodeList, err := cs.CoreV1().Nodes().List(metav1.ListOptions{FieldSelector: nodeSelector.String()})
+		if err != nil || len(nodeList.Items) != 1 {
+			framework.Failf("expected no err, got %v; expected len(nodes) = 1, got %v", err, len(nodeList.Items))
+		}
+		node := nodeList.Items[0]
+
+		By(fmt.Sprintf("Blocking traffic from node %s to the master", nodeName))
+		host, err := framework.GetNodeExternalIP(&node)
+		// TODO(Huang-Wei): make this case work for local provider
+		// if err != nil {
+		// 	host, err = framework.GetNodeInternalIP(&node)
+		// }
+		framework.ExpectNoError(err)
+		master := framework.GetMasterAddress(cs)
+		taint := newUnreachableNoExecuteTaint()
+
+		defer func() {
+			By(fmt.Sprintf("Unblocking traffic from node %s to the master", node.Name))
+			framework.UnblockNetwork(host, master)
+
+			if CurrentGinkgoTestDescription().Failed {
+				return
+			}
+
+			By(fmt.Sprintf("Expecting to see node %q becomes Ready", nodeName))
+			framework.WaitForNodeToBeReady(cs, nodeName, time.Minute*1)
+			By("Expecting to see unreachable=:NoExecute taint is taken off")
+			err := framework.WaitForNodeHasTaintOrNot(cs, nodeName, taint, false, time.Second*30)
+			framework.ExpectNoError(err)
+		}()
+
+		framework.BlockNetwork(host, master)
+
+		By(fmt.Sprintf("Expecting to see node %q becomes NotReady", nodeName))
+		if !framework.WaitForNodeToBeNotReady(cs, nodeName, time.Minute*3) {
+			framework.Failf("node %q doesn't turn to NotReady after 3 minutes", nodeName)
+		}
+		By("Expecting to see unreachable=:NoExecute taint is applied")
+		err = framework.WaitForNodeHasTaintOrNot(cs, nodeName, taint, true, time.Second*30)
+		framework.ExpectNoError(err)
+
+		By("Expecting pod0 to be evicted immediately")
+		err = framework.WaitForPodCondition(cs, ns, pods[0].Name, "pod0 terminating", time.Second*5, func(pod *v1.Pod) (bool, error) {
+			// as node is unreachable, pod0 is expected to be in Terminating status
+			// rather than getting deleted
+			if pod.DeletionTimestamp != nil {
+				return true, nil
+			}
+			return false, nil
+		})
+		framework.ExpectNoError(err)
+
+		By("Expecting pod2 to be updated with a toleration with tolerationSeconds=300")
+		err = framework.WaitForPodCondition(cs, ns, pods[2].Name, "pod2 updated with tolerationSeconds=300", time.Second*5, func(pod *v1.Pod) (bool, error) {
```

Comment:
@Huang-Wei - So, I am not sure, if the intention of the current PR is to check - if we are applying the given tolerations or expecting pod's status to be changed(meaning `pod.Status=='terminating' or 'running`) etc. I'd rather prefer the 2nd option, if possible. The way I am thinking about the test is to wait for pod status to change within the timeframe and see if the timeframe is greater than the toleration seconds that we have for it. Please let me know, if you think otherwise.

Reply:
You're right, it's using the 2nd option.

- pod0 has 0 seconds set, so it's expected to be in terminating status.
- pod1 has 200 seconds set, so nothing changed, but we don't want this test to wait for 200 seconds. So just check whether its seconds is unchanged.
- pod2 has no toleration set, so it's expected to be applied with a tolerationSeconds=300. But similar like pod1, we don't want this test to wait for 300 seconds. So still check tolerationsSeconds equals to 300 or not.

Reply:
What about the pod phase, is it still going to be running for pod1 and pod2? Are you not noticing any change in them?

Reply:
@ravisantoshgudimetla per our talk in slack, this "special" case (node unreachable), yes, pod1 and pod2 will be still in Running state.

Reply:
Ok then please add a check for pod.Status.Reason to be much more clear in cases where the pod status changes.